src/main/java/io/testproject/sdk/internal/helpers/ShutdownThreadManager.java

```diff
@@ -0,0 +1,128 @@
+/*
+ * Copyright (c) 2021 TestProject LTD. and/or its affiliates
+ * and other contributors as indicated by the @author tags.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.testproject.sdk.internal.helpers;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.Collection;
+import java.util.HashMap;
+
+/**
+ * Will be used to manage the driver and AgentClient shutdown threads to ensure
+ * the AgentClient will not close before the driver finishes reporting.
+ */
+public final class ShutdownThreadManager extends Thread {
+
+    /**
+     * Instance of the class.
+     */
+    private static ShutdownThreadManager instance;
+
+    /**
+     * Store the AgentClient shutdown thread.
+     */
+    private Runnable agentClientShutdownThread;
+
+    /**
+     * Hash map to store the shutdown threads of all the active drivers.
+     */
+    private HashMap<Integer, Runnable> driverShutdownThreads;
+
+    /**
+     * Logger instance.
+     */
+    private static final Logger LOG = LoggerFactory.getLogger(ShutdownThreadManager.class);
+
+    /**
+     * Add Agent Client shutdown thread to the manager.
+     *
+     * @param shutdownThread of the agent client.
+     */
+    public void addClientShutdownThread(final Runnable shutdownThread) {
```

Comment:
The class is called `ShutdownThreadManager`, so no reason adding `...ShutdownThread()` to all the methods.

```suggestion
    public void addAgentClient(final Runnable shutdownThread) {
```

Reply:
Makes sense, addressed.